units/github_status_api.go

```diff
@@ -0,0 +1,278 @@
+	tc := oauth2.NewClient(ctx, ts)
+	client := github.NewClient(tc)
+
+	newStatus := github.RepoStatus{
+		TargetURL:   github.String(fmt.Sprintf("%s%s", evergreenBaseURL, j.urlPath)),
+		Context:     github.String(j.context),
+		Description: github.String(j.description),
+		State:       github.String(j.ghStatus),
+	}
+	status, resp, err := client.Repositories.CreateStatus(ctx, j.owner, j.repo, j.ref, &newStatus)
+	if err != nil {
+		return err
+	}
+	resp.Body.Close()
+	if resp.StatusCode != http.StatusCreated {
+		repo := repoReference(j.owner, j.repo, j.prNumber, j.ref)
+		return errors.Errorf("Github status creation for %s: expected http Status code: 201 Created, got %s", repo, http.StatusText(http.StatusCreated))
+	}
+	if status == nil {
+		return errors.New("nil response from github")
+	}
+
+	return nil
+}
+
+func (j *githubStatusUpdateJob) fetch() error {
+	patchVersion := j.FetchID
+	if j.UpdateType == "build" {
+		b, err := build.FindOne(build.ById(j.FetchID))
+		if err != nil {
+			return err
+		}
+		if b == nil {
+			return errors.New("can't find build")
+		}
+
+		patchVersion = b.Version
+		j.context = fmt.Sprintf("evergreen-%s", b.BuildVariant)
+		j.description = taskStatusToDesc(b)
+		j.urlPath = fmt.Sprintf("/build/%s", b.Id)
+
+		switch b.Status {
+		case evergreen.BuildSucceeded:
+			j.ghStatus = githubStatusSuccess
+
+		case evergreen.BuildFailed:
+			j.ghStatus = githubStatusFailure
+
+		default:
+			return errors.New("build status is pending; refusing to update status")
+		}
+	}
+
+	patchDoc, err := patch.FindOne(patch.ByVersion(patchVersion))
+	if err != nil {
+		return err
+	}
+	if patchDoc == nil {
+		return errors.New("can't find patch")
+	}
+
+	if j.UpdateType == "patch" {
+		j.urlPath = fmt.Sprintf("/version/%s", patchVersion)
+		j.context = "evergreen"
+
+		switch patchDoc.Status {
+		case evergreen.PatchSucceeded:
+			j.ghStatus = githubStatusSuccess
+			j.description = fmt.Sprintf("finished in %s", patchDoc.FinishTime.Sub(patchDoc.StartTime).String())
+
+		case evergreen.PatchFailed:
+			j.ghStatus = githubStatusFailure
+			j.description = fmt.Sprintf("finished in %s", patchDoc.FinishTime.Sub(patchDoc.StartTime).String())
+
+		case evergreen.PatchCreated:
+			j.ghStatus = githubStatusPending
+			j.description = "preparing to run tasks"
+
+		case evergreen.PatchStarted:
+			j.ghStatus = githubStatusPending
+			j.description = "tasks are running"
+
+		default:
+			return errors.New("unknown patch status")
+		}
+	}
+
+	j.owner = patchDoc.GithubPatchData.BaseOwner
+	j.repo = patchDoc.GithubPatchData.BaseRepo
+	j.prNumber = patchDoc.GithubPatchData.PRNumber
+	j.ref = patchDoc.GithubPatchData.HeadHash
+
+	return nil
+}
+
+func (j *githubStatusUpdateJob) Run() {
+	if err := j.fetch(); err != nil {
+		j.AddError(err)
+		j.MarkComplete()
+		return
+	}
+
+	j.AddError(j.sendStatusUpdate())
+	if j.HasErrors() {
+		j.logger.Error(message.Fields{
+			"message": "github API failure",
+			"error":   j.Error(),
+		})
+		return
+	}
```

Comment:
I would log the errors as they happen, and if properly annotated, trust (as is generally true) that amboy will handle high level logging like this. 

Additionally by not marking complete here, you'll end up retrying (in most cases, very quickly.) 

I might do something like: comment on the github issue/send an alert/etc. that we had trouble contacting github. (grip has a sender for github comments) alerts already go to slack.

Reply:
If there is a Github API error, it's likely that we won't be able to post a message to the PR. A slack alert likely makes sense. Do I need to use a special sender, or will grip.Alert(...) work fine?

Reply:
Also, does amboy have a mechanism to say "retry, but after 60 minutes" or something to that affect?

Reply:
alas it doesn't, I think jobs should retry internally (e.g. retry requests to external sources,) but if after a reasonable period things just don't work, I think it's fine to give up and alert. 

Having said that, I might not make this an alert, as it's not particularly actionable on our end. 

Might it be useful to add to this message, something about the id of the patch/pull request that we're missing? That *might* make it more actionable for us.